packages/sqlite-pool/src/index.ts

```diff
@@ -0,0 +1,168 @@
+import createBaseConnectionPool, {
+  ConnectionPool,
+  PoolConnection,
+  PoolOptions,
+} from '@databases/connection-pool';
+import {once} from 'events';
+
+export type {SQLQuery};
+export {sql, isSqlQuery};
+
+export interface DatabaseTransaction {
+  query(query: SQLQuery): Promise<any[]>;
+
+  queryStream(query: SQLQuery): AsyncIterableIterator<any>;
+}
+
+export interface DatabaseConnection extends DatabaseTransaction {
+  tx<T>(fn: (db: DatabaseTransaction) => Promise<T>): Promise<T>;
+  dispose(): Promise<void>;
+}
+
+async function* transactionalQueryStream(
+  transaction: TransactionImplementation,
+  query: SQLQuery,
+): AsyncIterableIterator<any> {
+  const connection = transaction.connection;
+  for (const row of connection.queryStream(query)) {
+    if (transaction.aborted) {
+      throw new Error('Transaction aborted');
+    }
+    yield row;
+  }
+}
+
+class TransactionImplementation implements DatabaseTransaction {
+  connection: SyncDatabaseConnection;
+  aborted: boolean = false;
+
+  constructor(connection: SyncDatabaseConnection) {
+    this.connection = connection;
+  }
+
+  async query(query: SQLQuery): Promise<any[]> {
+    if (this.aborted) {
+      throw new Error('Transaction aborted');
+    }
+    return this.connection.query(query);
+  }
+
+  queryStream(query: SQLQuery): AsyncIterableIterator<any> {
+    return transactionalQueryStream(this, query);
+  }
+}
+
+async function* queryStream(
+  maybePoolConnection: Promise<
+    PoolConnection<SyncDatabaseConnectionWithController>
+  >,
+  query: SQLQuery,
+) {
+  const poolConnection = await maybePoolConnection;
+  try {
+    for (const row of poolConnection.connection.queryStream(query)) {
+      yield row;
+    }
+  } finally {
+    poolConnection.release();
+  }
+}
+
+type PartialPoolOptions = Omit<
+  PoolOptions<SyncDatabaseConnection>,
+  'openConnection' | 'closeConnection'
+>;
+
+interface SyncDatabaseConnectionWithController extends SyncDatabaseConnection {
+  controller?: AbortController;
+}
+
+class DatabaseConnectionImplementation implements DatabaseConnection {
+  #pool: ConnectionPool<SyncDatabaseConnectionWithController>;
+
+  constructor(
+    filename?: string,
+    options?: DatabaseOptions,
+    poolOptions?: PartialPoolOptions,
+  ) {
+    this.#pool = createBaseConnectionPool({
+      async openConnection() {
+        return connect(filename, options);
+      },
+      async closeConnection(connection) {
+        connection.dispose();
+        return;
+      },
+      async onReleaseTimeout(connection: SyncDatabaseConnectionWithController) {
+        const controller = connection.controller;
+        if (controller) {
+          controller.abort();
+        }
+        connection.dispose();
+        return;
+      },
+      ...poolOptions,
+    });
+  }
+
+  async query(query: SQLQuery): Promise<any[]> {
+    const poolConnection = await this.#pool.getConnection();
+    try {
+      const res = poolConnection.connection.query(query);
```

Comment:
Since this step, which runs the actual query, is synchronous, you cannot possibly get any parallel execution of queries with this approach.

To test this, you can run one very slow query (e.g. joining two very large tables) in parallel with one very fast query. If the slow query starts first, it will block the thread until it is finished, at which point the fast query will finally run.

To get actual parallel query execution using the synchronous SQLite client, you'll need to use worker threads (https://nodejs.org/api/worker_threads.html). You'd create one thread per connection and then pass queries to the threads to actually execute.

Reply:
Using worker_thread would be slower for the vast majority of cases unfortunately as the marshalling would wreck us. That's not the goal of what I'm trying to achieve with this anyway.

I have two goals:

1. achieve the same interface of the pg/mysql/sqlite modules without having a mutex
2. perform parallel read transactions.

Reply:
Right, you cannot have:

1. Synchronous underlying database client/connection
2. Single threaded
3. Parallel queries

You must choose two. As soon as you submit one query via `poolConnection.connection.query(query)`, the JavaScript thread that submitted the query is blocked until it receives a response, meaning that it cannot submit a second query to run in parallel, even if it has a separate connection to the database to run that parallel query on.

To demonstrate this, please create a file called `index.js` with the following contents:

```javascript
function createFakeConnection() {
  return {
    query: async () => {
      const end = Date.now() + 1_000;
      while (Date.now() < end);
    },
  };
}

async function run() {
  const connectionCount = parseInt(process.argv[2], 10);

  console.log(`Running ${connectionCount} parallel queries...`);

  const connections = [];
  for (let i = 0; i < connectionCount; i++) {
    connections.push(createFakeConnection());
  }

  const start = Date.now();
  await Promise.all(connections.map((connection) => connection.query()));
  const end = Date.now();

  console.log(`Took ${end - start}ms`);
}

run().catch((ex) => {
  console.error(ex);
  process.exit(1);
});
```

Now try running the following:

- `node index.js 1` - will take 1000ms
- `node index.js 2` - will take 2000ms
- `node index.js 3` - will take 3000ms

The `async` keyword does not create a thread, it just says "this function should return a promise and it may use the `await` keyword to yield to promises during execution". If the underlying resource is still synchronous, it will still block the entire thread. I've made the underlying resource a JavaScript `while` loop here to try and make this more obvious, but it makes no difference if it's in JavaScript or a native C++ module.

Getting back to the three options that you can only pick 2 from:

1. Synchronous underlying database client/connection
2. Single threaded
3. Parallel queries

We currently have:

 - `@databases/sqlite-sync` - chooses 1 and 2, but not 3 (i.e. no parallel queries)
 - `@databases/sqlite` - chooses 2 and 3, but not 1 (i.e. parallel read queries, but the underlying database client is async using threads within C++)

This proposal is still 1 and 2 but not 3, it's just hidden behind additional layers of complexity.

I agree that doing 1 and 3 may not be faster for a lot of use cases - It'll only help if the queries are slow but the amount of data is small so marshalling is fast. The only alternative though is to continue to use `sqlite3` which is actually async, or build something new at the native level that combines the async support of `sqlite3` with the better support for BIGINT etc from `better-sqlite3`.

`better-sqlite3` do propose & have a somewhat decent example of how to do the multi-threaded approach here: https://github.com/WiseLibs/better-sqlite3/blob/HEAD/docs/threads.md You'd need to do some benchmarking to see how much slower this made things.

Reply:
P.S. "achieve the same interface of the pg/mysql/sqlite modules" is fairly trivial, it just doesn't achieve the second goal of parallel queries.

Reply:
P.S. You can minimise somewhat the impact of marshalling by dealing with the `SQLQuery` package steps in the primary thread so that all that's passed through to the child threads is a `string` representation of the query and an array of values.